ui/StatusQ/src/writableproxymodel.cpp

```diff
@@ -0,0 +1,743 @@
+    auto itemData = q.sourceModel()->itemData(sourceIndex);
+    auto proxyIndex = q.mapFromSource(sourceIndex);
+    for(auto it = itemData.begin(); it != itemData.end(); ++it)
+    {
+        if(insertedRows[proxyIndex].contains(it.key()))
+            continue;
+
+        insertedRows[proxyIndex][it.key()] = it.value();
+    }
+}
+
+void WritableProxyModelPrivate::applyRoleNamesFix()
+{
+    if(!q.sourceModel())
+        return;
+    
+    if(q.sourceModel()->rowCount() != 0)
+        return;
+
+    auto connection = std::make_unique<QMetaObject::Connection>();
+    auto connectionPtr = connection.get();
+    
+    auto singleShot = [this, connection = std::move(connection)]() {
+        QObject::disconnect(*connection);
+        q.resetInternalData();
+    };
+    
+    *connectionPtr = q.connect(q.sourceModel(), &QAbstractItemModel::rowsInserted, &q, std::move(singleShot));
+}
+
+QVector<QPair<int, int>> WritableProxyModelPrivate::sourceRowRangesBetween(int start, int end) const
+{
+    QVector<QPair<int, int>> result;
+    int currentStart = -1;
+    int currentEnd = -1;
+    for(int i = start; i <= end; ++i)
+    {
+        auto proxyRow = sourceToProxyRow(i);
+        if(proxyRow >= 0)
+        {
+            if(currentStart == -1)
+            {
+                currentStart = proxyRow;
+                currentEnd = currentStart;
+            }
+            else
+            {
+                currentEnd = proxyRow;
+            }
+        }
+        else
+        {
+            if(currentStart != -1)
+            {
+                result.append({ currentStart, currentEnd });
+                currentStart = -1;
+                currentEnd = -1;
+            }
+        }
+    }
+
+    if(currentStart != -1)
+    {
+        result.append({ currentStart, currentEnd });
+    }
+
+    return result;
+}
+
+void WritableProxyModelPrivate::adjustInsertedRowsBy(int start, int offset)
+{
+    IndexedValues<QModelIndex> newInsertedRows;
+    for(auto iter = insertedRows.begin(); iter != insertedRows.end(); ++iter)
+    {
+        auto key = iter.key();
+        auto value = iter.value();
+        if(key.row() >= start)
+        {
+            newInsertedRows.insert(key.siblingAtRow(key.row() + offset), value);
+        }
+        else
+        {
+            newInsertedRows.insert(key, value);
+        }
+    }
+    insertedRows.swap(newInsertedRows);
+}
+
+WritableProxyModel::WritableProxyModel(QObject* parent)
+    : QAbstractProxyModel(parent)
+    , d(new WritableProxyModelPrivate(*this))
+{
+#ifdef QT_DEBUG
+    // Enable the model tester on debug builds to catch any model errors
+    new QAbstractItemModelTester(this, QAbstractItemModelTester::FailureReportingMode::Warning, this);
+#endif
+}
+
+WritableProxyModel::~WritableProxyModel() = default;
+
+QVariantMap WritableProxyModel::toVariantMap() const
+{
+    QVariantMap result;
+    for(int row = 0; row < rowCount(); ++row)
```

Comment:
You definitely want to move the `rowCount()` call out of the loop here

Reply:
Good point! Fixed. `roleNames()` usage as well